api/src/main/java/org/openmrs/api/db/hibernate/HibernateConceptDAO.java

```diff
@@ -2068,4 +2043,43 @@ private Criteria createSearchDrugByMappingCriteria(String code, ConceptSource co
 		}
 		return searchCriteria;
 	}
+
+	private Criteria createSearchConceptMapCriteria(String code, String sourceName, boolean includeRetired) {
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(ConceptMap.class);
+
+		//join to the conceptReferenceTerm table
+		criteria.createAlias("conceptReferenceTerm", "term");
+
+		// match the source code to the passed code
+		if (Context.getAdministrationService().isDatabaseStringComparisonCaseSensitive()) {
+			criteria.add(Restrictions.eq("term.code", code).ignoreCase());
+		} else {
+			criteria.add(Restrictions.eq("term.code", code));
+		}
+
+		// join to concept reference source and match to the h17Code or source name
+		criteria.createAlias("term.conceptSource", "source");
+		if (Context.getAdministrationService().isDatabaseStringComparisonCaseSensitive()) {
+			criteria.add(Restrictions.or(Restrictions.eq("source.name", sourceName).ignoreCase(), Restrictions.eq(
+					"source.hl7Code", sourceName).ignoreCase()));
+		} else {
+			criteria.add(Restrictions.or(Restrictions.eq("source.name", sourceName), Restrictions.eq("source.hl7Code",
+					sourceName)));
+		}
+
+		criteria.createAlias("concept", "concept");
+
+		if (!includeRetired) {
+			// ignore retired concepts
+			criteria.add(Restrictions.eq("concept.retired", false));
```

Comment:
Note that cache should be evicted also when concept.retired changes.

Reply:
I just pushed a follow-up commit that:
* Deprecates the getConceptsByMapping DAO method in favor of future use of getConceptIdsByMapping
* Evicts from cache more liberally on saveConcept, purge/save/retire ConceptSource, updateConceptIndexes
Thanks for the support and feedback @rkorytkowski - much appreciated.